File: scraper/parser.py
import re
from dataclasses import dataclass
from bs4 import BeautifulSoup, Tag

GID_TOKEN_RE = re.compile(r"/g/(\d+)/([a-f0-9]+)/")
NEXT_CURSOR_RE = re.compile(r'[?&]next=(\d+)')
RATING_RE = re.compile(r"([0-5](?:\.\d+)?)")
TAG_CLASS_RE = re.compile(r"^gt")
OPACITY_RE = re.compile(r"opacity\s*:\s*([0-9.]+)")
SPACE_RE = re.compile(r"\s+")


@dataclass(frozen=True)
class GalleryListItem:
    gid: int
    token: str
    title: str
    rating_sig: str
    rating_est: float | None
    visible_tag_count: int


def _normalize_text(value: str) -> str:
    return SPACE_RE.sub(" ", value or "").strip()


def _extract_rating_signal(element: Tag) -> tuple[str, float | None]:
    ir_node = element.find(class_=re.compile(r"\bir"))
    if ir_node:
        title = _normalize_text(ir_node.get("title", ""))
        if title:
            m = RATING_RE.search(title)
            if m:
                value = float(m.group(1))
                return f"avg:{value:.2f}", value
            return f"title:{title}", None

        style = _normalize_text(ir_node.get("style", ""))
        if style:
            m = OPACITY_RE.search(style)
            if m:
                try:
                    opacity = float(m.group(1))
                    # EX list stars often encode fill ratio in opacity; map 0..1 to 0..5.
                    estimate = max(0.0, min(5.0, opacity * 5.0))
                    return f"style:{style.replace(' ', '')}", estimate
                except ValueError:
                    pass
            return f"style:{style.replace(' ', '')}", None

        classes = [c for c in (ir_node.get("class") or []) if c]
        if classes:
            return f"class:{','.join(sorted(classes))}", None

    for klass in ("gl4e", "gl4t", "gl5t", "gl5m", "gl5c"):
        node = element.find(class_=klass)
        if not node:
            continue
        text = _normalize_text(node.get_text(" ", strip=True))
        m = RATING_RE.search(text)
        if m:
            value = float(m.group(1))
            return f"text:{value:.2f}", value

    return "", None


def _extract_visible_tags(element: Tag) -> set[str]:
    tags: set[str] = set()

    for node in element.find_all(True):
        classes = node.get("class") or []
        if not any(TAG_CLASS_RE.match(c) for c in classes):
            continue
        text = _normalize_text(node.get_text(" ", strip=True)).lower()
        if not text:
            continue
        if len(text) > 80:
            continue
        tags.add(text)

    if tags:
        return tags

    for a in element.find_all("a", href=True):
        href = a.get("href", "")
        if "f_search=" not in href:
            continue
        text = _normalize_text(a.get_text(" ", strip=True)).lower()
        if not text:
            continue
        if len(text) > 80:
            continue
        if text in {"archive download", "torrent download"}:
            continue
        tags.add(text)
    return tags


def parse_gallery_list(html: str) -> tuple[list[GalleryListItem], int | None]:
    """
    解析列表页 HTML
    返回 (items, next_gid)
      items    = [GalleryListItem(...), ...]
      next_gid = 下一页游标 GID，None 表示已到最后一页
    """
    soup = BeautifulSoup(html, "lxml")
    results: list[GalleryListItem] = []

    itg = soup.find(class_="itg")
    if not itg:
        return results, None

    for element in itg.find_all(recursive=False):
        glname = element.find(class_="glname")
        if not glname:
            continue

        # 找到 <a> 链接
        a = glname.find("a")
        if a is None:
            parent = glname.parent
            if parent and parent.name == "a":
                a = parent
        if a is None:
            continue

        href = a.get("href", "")
        m = GID_TOKEN_RE.search(href)
        if not m:
            continue
        gid, token = int(m.group(1)), m.group(2)

        # 取最深子节点的文本作为标题
        node = glname
        while True:
            tag_children = [c for c in node.children if isinstance(c, Tag)]
            if not tag_children:
                break
            node = tag_children[0]
        title = _normalize_text(node.get_text(" ", strip=True) if isinstance(node, Tag) else str(node))
        rating_sig, rating_est = _extract_rating_signal(element)
        visible_tags = _extract_visible_tags(element)

        results.append(
            GalleryListItem(
                gid=gid,
                token=token,
                title=title,
                rating_sig=rating_sig,
                rating_est=rating_est,
                visible_tag_count=len(visible_tags),
            )
        )

    # 从分页栏提取下一页游标
    # ExHentai 分页: id="dnext" 为"下一页"按钮，href 含 next=<gid>
    next_gid: int | None = None
    dnext = soup.find(id="dnext")
    if dnext:
        href = dnext.get("href", "")
        m = NEXT_CURSOR_RE.search(href)
        if m:
            next_gid = int(m.group(1))

    return results, next_gid


def parse_detail(html: str) -> dict:
    """
    解析画廊详情页 HTML，返回字段 dict。
    """
    soup = BeautifulSoup(html, "lxml")
    data: dict = {}

    gm = soup.find(class_="gm")
    if not gm:
        return data

    # 标题
    gn = gm.find(id="gn")
    data["title"] = gn.get_text(strip=True) if gn else ""
    gj = gm.find(id="gj")
    data["title_jpn"] = gj.get_text(strip=True) if gj else ""

    # 分类  .cn（彩色）或 .cs（灰色）
    ce = gm.find(class_="cn") or gm.find(class_="cs")
    data["category"] = ce.get_text(strip=True) if ce else ""

    # 上传者
    gdn = gm.find(id="gdn")
    data["uploader"] = gdn.get_text(strip=True) if gdn else ""

    # 评分
    rating_label = gm.find(id="rating_label")
    if rating_label:
        rtext = rating_label.get_text(strip=True)
        if "Not Yet Rated" not in rtext:
            idx = rtext.find(" ")
            if idx != -1:
                try:
                    data["rating"] = float(rtext[idx + 1:])
                except ValueError:
                    pass
    
    # 封面 URL
    gd1 = gm.find(id="gd1")
    if gd1:
        div = gd1.find("div")
        if div:
            style = div.get("style", "")
            m = re.search(r"url\((.+?)\)", style)
            if m:
                data["thumb"] = m.group(1).strip("'\"")

    # 详情表格 #gdd
    gdd = gm.find(id="gdd")
    if gdd:
        for tr in gdd.find_all("tr"):
            tds = tr.find_all("td")
            if len(tds) < 2:
                continue
            key = tds[0].get_text(strip=True)
            value = tds[1].get_text(strip=True)

            if key.startswith("Posted"):
                data["posted"] = value
            elif key.startswith("Language"):
                data["language"] = value
            elif key.startswith("Length"):
                idx = value.find(" ")
                if idx >= 0:
                    try:
                        data["pages"] = int(value[:idx].replace(",", ""))
                    except ValueError:
                        pass
            elif key.startswith("Favorited"):
                if value == "Never":
                    data["fav_count"] = 0
                elif value == "Once":
                    data["fav_count"] = 1
                else:
                    idx = value.find(" ")
                    if idx >= 0:
                        try:
                            data["fav_count"] = int(value[:idx].replace(",", ""))
                        except ValueError:
                            data["fav_count"] = 0
                            
    # Comment count
    cdiv = soup.find(id="cdiv")
    if cdiv:
        aall = cdiv.find(id="aall")
        if aall:
            m = re.search(r"(\d+)", aall.get_text())
            data["comment_count"] = int(m.group(1)) if m else len(cdiv.find_all(class_="c1"))
        else:
            data["comment_count"] = len(cdiv.find_all(class_="c1"))
    else:
        data["comment_count"] = 0

    # Tags extraction
    taglist = soup.find(id="taglist")
    if taglist:
        tags = {}
        # Each row in table inside taglist represents a namespace? Not always.
        # Actually structure is usually <table><tr><td>namespace:</td><td><div><a>tag</a></div>...</td></tr>...</table>
        for tr in taglist.find_all("tr"):
            tds = tr.find_all("td")
            if len(tds) < 2:
                continue
            # namespace like 'language:', 'female:', or empty for misc
            ns_text = tds[0].get_text(strip=True).rstrip(":")
            namespace = ns_text if ns_text else "misc"
            
            tag_values = []
            for div in tds[1].find_all("div"):
                a = div.find("a")
                if a:
                    t = a.get_text(strip=True)
                    # sometimes tags have space or _
                    tag_values.append(t)
            
            if tag_values:
                tags[namespace] = tag_values
        data["tags"] = tags

    return data

File: scraper-go/parser/parser.go
package parser

import (
	"regexp"
	"strconv"
	"strings"

	"github.com/PuerkitoBio/goquery"
)

var (
	gidTokenRE   = regexp.MustCompile(`/g/(\d+)/([a-f0-9]+)/`)
	nextCursorRE = regexp.MustCompile(`[?&]next=([^&"\s]+)`)
	ratingRE     = regexp.MustCompile(`([0-5](?:\.\d+)?)`)
	totalCountRE = regexp.MustCompile(`Found\s+(?:about\s+)?([\d,]+)\s+results`)
	tagClassRE   = regexp.MustCompile(`^gt`)
	bgPosRE      = regexp.MustCompile(`background-position\s*:\s*(-?\d+)px\s+(-?\d+)px`)
	spaceRE      = regexp.MustCompile(`\s+`)
	thumbURLRE   = regexp.MustCompile(`url\((.+?)\)`)
	postedIDRE   = regexp.MustCompile(`^posted_`)
	digitRE      = regexp.MustCompile(`(\d+)`)
)

type GalleryListItem struct {
	GID         int64
	Token       string
	Title       string
	RatingSig   string
	RatingEst   *float64
	VisibleTags []string
	FavoritedAt *string
	IsDeleted   bool
}

type ListResult struct {
	Items      []GalleryListItem
	NextCursor *string
	TotalCount *int
}

type GalleryDetail struct {
	Title        string
	TitleJPN     string
	Category     string
	Uploader     string
	Rating       *float64
	Thumb        string
	Posted       string
	Language     string
	Pages        int
	FavCount     int
	CommentCount int
	Tags         map[string][]string
}

func normalizeText(s string) string {
	return strings.TrimSpace(spaceRE.ReplaceAllString(s, " "))
}

func extractRatingSignal(sel *goquery.Selection) (string, *float64) {
	// Try CSS sprite-based rating
	sel.Find(".ir").Each(func(_ int, ir *goquery.Selection) {
		// Already found? skip via closure
	})

	var sig string
	var est *float64

	sel.Find("[class*='ir']").EachWithBreak(func(_ int, ir *goquery.Selection) bool {
		style, _ := ir.Attr("style")
		style = normalizeText(style)
		m := bgPosRE.FindStringSubmatch(style)
		if m == nil {
			title, _ := ir.Attr("title")
			title = normalizeText(title)
			if title == "" {
				return true // continue
			}
			mt := ratingRE.FindStringSubmatch(title)
			if mt != nil {
				v, _ := strconv.ParseFloat(mt[1], 64)
				sig = "title:" + mt[1]
				est = &v
				return false // break
			}
			return true
		}
		x, _ := strconv.Atoi(m[1])
		y, _ := strconv.Atoi(m[2])
		if y == -1 {
			v := 5.0 - float64(abs(x))/16.0
			v = clamp(v, 0, 5)
			sig = "sprite:x=" + m[1] + ",y=" + m[2]
			est = &v
			return false
		}
		if y == -21 {
			v := 4.5 - float64(abs(x))/16.0
			v = clamp(v, 0, 5)
			sig = "sprite:x=" + m[1] + ",y=" + m[2]
			est = &v
			return false
		}
		return true
	})

	if est != nil {
		return sig, est
	}

	// Fallback: text-based rating
	for _, class := range []string{".gl4e", ".gl4t", ".gl5t", ".gl5m", ".gl5c"} {
		node := sel.Find(class)
		if node.Length() == 0 {
			continue
		}
		text := normalizeText(node.Text())
		m := ratingRE.FindStringSubmatch(text)
		if m != nil {
			v, _ := strconv.ParseFloat(m[1], 64)
			sig = "text:" + m[1]
			est = &v
			return sig, est
		}
	}
	return "", nil
}

func extractVisibleTags(sel *goquery.Selection) []string {
	tags := make(map[string]struct{})

	sel.Find("*").Each(func(_ int, node *goquery.Selection) {
		classes, _ := node.Attr("class")
		if classes == "" {
			return
		}
		for _, c := range strings.Fields(classes) {
			if tagClassRE.MatchString(c) {
				text := strings.ToLower(normalizeText(node.Text()))
				if text != "" && len(text) <= 80 {
					tags[text] = struct{}{}
				}
				return
			}
		}
	})

	if len(tags) > 0 {
		return mapKeys(tags)
	}

	// Fallback: f_search links
	sel.Find("a[href]").Each(func(_ int, a *goquery.Selection) {
		href, _ := a.Attr("href")
		if !strings.Contains(href, "f_search=") {
			return
		}
		text := strings.ToLower(normalizeText(a.Text()))
		if text == "" || len(text) > 80 {
			return
		}
		if text == "archive download" || text == "torrent download" {
			return
		}
		tags[text] = struct{}{}
	})
	return mapKeys(tags)
}

func ParseGalleryList(html string) (*ListResult, error) {
	doc, err := goquery.NewDocumentFromReader(strings.NewReader(html))
	if err != nil {
		return nil, err
	}

	result := &ListResult{}

	// Total count
	if m := totalCountRE.FindStringSubmatch(html); m != nil {
		n, _ := strconv.Atoi(strings.ReplaceAll(m[1], ",", ""))
		result.TotalCount = &n
	}

	itg := doc.Find(".itg")
	if itg.Length() == 0 {
		return result, nil
	}

	// In extended mode (.itg is a <table>), iterate over <tr> rows.
	// In other modes (.itg is a <div>), iterate over direct children.
	var rows *goquery.Selection
	if goquery.NodeName(itg) == "table" {
		rows = itg.Find("tr")
	} else {
		rows = itg.Children()
	}

	rows.Each(func(_ int, el *goquery.Selection) {
		glname := el.Find(".glname")
		if glname.Length() == 0 {
			return
		}

		// Find <a> link
		a := glname.Find("a").First()
		if a.Length() == 0 {
			// Check parent
			parent := glname.Parent()
			if goquery.NodeName(parent) == "a" {
				a = parent
			}
		}
		if a.Length() == 0 {
			return
		}

		href, _ := a.Attr("href")
		m := gidTokenRE.FindStringSubmatch(href)
		if m == nil {
			return
		}
		gid, _ := strconv.ParseInt(m[1], 10, 64)
		token := m[2]

		// Title: deepest child text
		title := extractDeepestText(glname)
		ratingSig, ratingEst := extractRatingSignal(el)
		visibleTags := extractVisibleTags(el)

		// Favorited at
		var favAt *string
		el.Find("p").EachWithBreak(func(_ int, p *goquery.Selection) bool {
			if strings.TrimSpace(p.Text()) == "Favorited:" {
				next := p.Next()
				if next.Length() > 0 {
					text := strings.TrimSpace(next.Text())
					if text != "" {
						favAt = &text
					}
				}
				return false
			}
			return true
		})

		// Deleted detection
		isDeleted := false
		el.Find("[id]").Each(func(_ int, node *goquery.Selection) {
			id, _ := node.Attr("id")
			if postedIDRE.MatchString(id) && node.Find("s").Length() > 0 {
				isDeleted = true
			}
		})

		result.Items = append(result.Items, GalleryListItem{
			GID:         gid,
			Token:       token,
			Title:       title,
			RatingSig:   ratingSig,
			RatingEst:   ratingEst,
			VisibleTags: visibleTags,
			FavoritedAt: favAt,
			IsDeleted:   isDeleted,
		})
	})

	// Next cursor
	dnext := doc.Find("#dnext")
	if dnext.Length() > 0 {
		href, _ := dnext.Attr("href")
		m := nextCursorRE.FindStringSubmatch(href)
		if m != nil {
			result.NextCursor = &m[1]
		}
	}

	return result, nil
}

func ParseDetail(html string) (*GalleryDetail, error) {
	doc, err := goquery.NewDocumentFromReader(strings.NewReader(html))
	if err != nil {
		return nil, err
	}

	gm := doc.Find(".gm")
	if gm.Length() == 0 {
		return nil, nil
	}

	d := &GalleryDetail{}

	d.Title = strings.TrimSpace(gm.Find("#gn").Text())
	d.TitleJPN = strings.TrimSpace(gm.Find("#gj").Text())

	// Category
	ce := gm.Find(".cn")
	if ce.Length() == 0 {
		ce = gm.Find(".cs")
	}
	d.Category = strings.TrimSpace(ce.Text())

	// Uploader
	d.Uploader = strings.TrimSpace(gm.Find("#gdn").Text())

	// Rating
	ratingLabel := gm.Find("#rating_label")
	if ratingLabel.Length() > 0 {
		rtext := strings.TrimSpace(ratingLabel.Text())
		if !strings.Contains(rtext, "Not Yet Rated") {
			idx := strings.Index(rtext, " ")
			if idx != -1 {
				if v, err := strconv.ParseFloat(rtext[idx+1:], 64); err == nil {
					d.Rating = &v
				}
			}
		}
	}

	// Thumb URL
	gd1 := gm.Find("#gd1 div")
	if gd1.Length() > 0 {
		style, _ := gd1.Attr("style")
		if m := thumbURLRE.FindStringSubmatch(style); m != nil {
			d.Thumb = strings.Trim(m[1], "'\"")
		}
	}

	// Detail table #gdd
	gm.Find("#gdd tr").Each(func(_ int, tr *goquery.Selection) {
		tds := tr.Find("td")
		if tds.Length() < 2 {
			return
		}
		key := strings.TrimSpace(tds.First().Text())
		value := strings.TrimSpace(tds.Last().Text())

		switch {
		case strings.HasPrefix(key, "Posted"):
			d.Posted = value
		case strings.HasPrefix(key, "Language"):
			d.Language = value
		case strings.HasPrefix(key, "Length"):
			idx := strings.Index(value, " ")
			if idx >= 0 {
				if n, err := strconv.Atoi(strings.ReplaceAll(value[:idx], ",", "")); err == nil {
					d.Pages = n
				}
			}
		case strings.HasPrefix(key, "Favorited"):
			switch {
			case value == "Never":
				d.FavCount = 0
			case value == "Once":
				d.FavCount = 1
			default:
				idx := strings.Index(value, " ")
				if idx >= 0 {
					if n, err := strconv.Atoi(strings.ReplaceAll(value[:idx], ",", "")); err == nil {
						d.FavCount = n
					}
				}
			}
		}
	})

	// Comment count
	cdiv := doc.Find("#cdiv")
	if cdiv.Length() > 0 {
		aall := cdiv.Find("#aall")
		if aall.Length() > 0 {
			if m := digitRE.FindStringSubmatch(aall.Text()); m != nil {
				d.CommentCount, _ = strconv.Atoi(m[1])
			} else {
				d.CommentCount = cdiv.Find(".c1").Length()
			}
		} else {
			d.CommentCount = cdiv.Find(".c1").Length()
		}
	}

	// Tags
	taglist := doc.Find("#taglist")
	if taglist.Length() > 0 {
		d.Tags = make(map[string][]string)
		taglist.Find("tr").Each(func(_ int, tr *goquery.Selection) {
			tds := tr.Find("td")
			if tds.Length() < 2 {
				return
			}
			ns := strings.TrimRight(strings.TrimSpace(tds.First().Text()), ":")
			if ns == "" {
				ns = "misc"
			}
			var tagValues []string
			tds.Last().Find("div a").Each(func(_ int, a *goquery.Selection) {
				t := strings.TrimSpace(a.Text())
				if t != "" {
					tagValues = append(tagValues, t)
				}
			})
			if len(tagValues) > 0 {
				d.Tags[ns] = tagValues
			}
		})
	}

	return d, nil
}

func extractDeepestText(sel *goquery.Selection) string {
	node := sel
	for {
		children := node.Children()
		if children.Length() == 0 {
			break
		}
		node = children.First()
	}
	return normalizeText(node.Text())
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func clamp(v, lo, hi float64) float64 {
	if v < lo {
		return lo
	}
	if v > hi {
		return hi
	}
	return v
}

func mapKeys(m map[string]struct{}) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	return keys
}
